File: leetcode/9-palindrome-number/main.go
package main

import (
	"fmt"
	"math"
)

func isPalindrome(x int) bool {
	if x <= 0 {
		return x == 0
	}

	length := int(math.Log10(float64(x)))
	mask := int(math.Pow10(length))
	for i := 0; i <= length/2; i++ {
		if x%10 != x/mask {
			return false
		}
		x %= mask
		x /= 10
		mask /= 100
	}

	return true
}

/*func isPalindrome(x int) bool {
	if x < 0 {
		return false
	}

	tmp := x
	var reversed int
	for tmp > 0 {
		reversed = reversed*10 + tmp%10
		tmp /= 10
	}

	return reversed == x
}*/

func main() {
	x := 12321
	fmt.Println(isPalindrome(x))
}

File: leetcode/9-palindrome-number/main.py
import math


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x <= 0:
            return x == 0

        digits = math.floor(math.log10(x)) + 1
        mask = 10 ** (digits - 1)
        for i in range(digits // 2):
            if x % 10 != x // mask:
                return False
            x %= mask
            x //= 10
            mask //= 100

        return True


solution = Solution()
print(solution.isPalindrome(12322))
